lib/src/util/condition_scope_visitor.dart

```diff
@@ -0,0 +1,259 @@
+    node.visitChildren(this);
+  }
+
+  void visitCondition(Expression node);
+
+  @override
+  visitDoStatement(DoStatement node) {
+    visitCondition(node.condition);
+    bool hasBreak = DartTypeUtilities
+        .traverseNodesInDFS(node, excludeCriteria: (node) => _isLoopStatement(node))
+        .any((e) => e is BreakStatement);
+    if (!hasBreak) {
+      _addElementToEnvironment(
+          new _ConditionExpression(node.condition, value: false));
+    }
+  }
+
+  @override
+  visitForStatement(ForStatement node) {
+    _addLocalEnvironment();
+    _addElementToEnvironment(new _ConditionExpression(node.condition));
+    node.variables?.accept(this);
+    node.initialization?.accept(this);
+    visitCondition(node.condition);
+    node.condition?.accept(this);
+    _addElementToEnvironment(new _ConditionExpression(node.condition));
+    node.updaters.accept(this);
+    node.body?.accept(this);
+    _removeLocalEnvironment();
+    bool hasBreak = DartTypeUtilities
+        .traverseNodesInDFS(node, excludeCriteria: _isLoopStatement)
+        .any((e) => e is BreakStatement);
+    if (!hasBreak) {
+      _addElementToEnvironment(
+          new _ConditionExpression(node.condition, value: false));
+    }
+  }
+
+  @override
+  visitIfStatement(IfStatement node) {
+    visitCondition(node.condition);
+    final elseUndefinedExpressions =
+        _visitElseStatement(node.elseStatement, node.condition);
+    _visitIfStatement(node);
+    elseUndefinedExpressions
+        .where((e) => e is _UndefinedExpression)
+        .forEach(_addElementToEnvironment);
+    if (_isLastStatementAnExitStatement(node.thenStatement)) {
+      _addElementToEnvironment(
+          new _ConditionExpression(node.condition, value: false));
+    }
+    if (_isLastStatementAnExitStatement(node.elseStatement)) {
+      _addElementToEnvironment(new _ConditionExpression(node.condition));
+    }
+  }
+
+  @override
+  visitNode(AstNode node) {
+    final needsLocalEnvironment = _needsLocalEnvironment(node);
+    if (needsLocalEnvironment) {
+      _addLocalEnvironment();
+      _addElementToEnvironment(new _UndefinedAllExpression());
+    }
+    node.visitChildren(this);
+    if (needsLocalEnvironment) {
+      // do not need to propagate undefined variables.
+      environments.removeFirst();
+    }
+  }
+
+  @override
+  visitPostfixExpression(PostfixExpression node) {
+    final operand = node.operand;
+    if (operand is SimpleIdentifier) {
+      _addElementToEnvironment(new _UndefinedExpression(operand.bestElement));
+    }
+    node.visitChildren(this);
+  }
+
+  @override
+  visitPrefixExpression(PrefixExpression node) {
+    final operand = node.operand;
+    if (operand is SimpleIdentifier) {
+      _addElementToEnvironment(new _UndefinedExpression(operand.bestElement));
+    }
+    node.visitChildren(this);
+  }
+
+  @override
+  visitVariableDeclaration(VariableDeclaration node) {
+    _addElementToEnvironment(new _UndefinedExpression(node.element));
```

Comment:
It is not supported by now, and before this optimization it was not contemplated, but it could be implemented in the future, I would like to do it, what do you think @alexeieleusis ?